<?xml version="1.0" encoding="UTF-8" ?>
<launch>
    <arg name="device" default="cuda" doc="Device to run tensor operations on: cpu or cuda"/>
    <arg name="cloud" default="points" doc="Input point cloud for segmentation"/>
    <arg name="max_age" default="0.5"/>
    <arg name="debug" default="false"/>
    <arg name="weights" default="deeplabv3_resnet101_lr_0.0001_bs_8_epoch_90_TraversabilityClouds_depth_labels_traversability_iou_0.972.pth"/>
    <arg name="height" default="128"/>
    <arg name="width" default="1024"/>
    <arg name="fov_elevation" default="90"/>  <!-- 90 deg -->

<!--    Point cloud segmentation -->
    <node name="cloud_segmentation" pkg="traversability_estimation" type="cloud_segmentation_rgb">
        <rosparam subst_value="true">
            device: $(arg device)
            max_age: $(arg max_age)
            lidar_channels: $(arg height)
            lidar_beams: $(arg width)
            range_projection: true
            debug: $(arg debug)
            soft_label_ind: 1
            weights: $(arg weights)
        </rosparam>
        <param name="lidar_fov_up" value="$(eval arg('fov_elevation') / 2.)"/>
        <param name="lidar_fov_down" value="$(eval -arg('fov_elevation') / 2.)"/>

        <remap from="cloud_in" to="$(arg cloud)"/>
        <remap from="cloud_out" to="cloud_segmentation/points"/>
    </node>

<!--    Cloud voxel filtering: 10 cm -->
    <node name="cloud_voxel_filter_10cm" pkg="nodelet" type="nodelet"
          args="standalone cloud_proc/voxel_filter"
            respawn="true" respawn_delay="1.0" output="log">
        <rosparam>
            field: x
            grid: 0.1
            zero_valid: false
        </rosparam>
        <remap from="input" to="cloud_segmentation/points"/>
        <remap from="output" to="cloud_segmentation/points_10cm"/>
    </node>
<!--    Cloud voxel filtering: 20 cm -->
    <node name="cloud_voxel_filter_20cm" pkg="nodelet" type="nodelet"
          args="standalone cloud_proc/voxel_filter"
            respawn="true" respawn_delay="1.0" output="log">
        <rosparam>
            field: x
            grid: 0.2
            zero_valid: false
        </rosparam>
        <remap from="input" to="cloud_segmentation/points_10cm"/>
        <remap from="output" to="cloud_segmentation/points_20cm"/>
    </node>

</launch>
